File: 0343-integer-break/0343-integer-break.py
class Solution:
    def integerBreak(self, n: int) -> int:
        @lru_cache(None)
        def compute(n):
            if n==1 or n==2:
                return 1
            res = 1
            for i in range(1,n):
                v1 = compute(n-i)
                v2 = compute(i)
                res = max(res,v1*v2)
                res = max(res,i*(n-i))
                res = max(res,v1 * (i))
                res = max(res,v2 * (n-i) )
            return res
        return compute(n)

File: 0343-integer-break/0343-integer-break.java
class Solution {
    private Integer[] dp ;
    public int compute(int num){
        if(num<=2){
            return 1;
        }
        if(dp[num]!=null) return dp[num];
        int res =1;
        for(int i = 1 ; i<num ; i++){
            int v1 = compute(i);
            int v2 = compute(num-i);
            res = Math.max(res,v1*v2);
            res = Math.max(res,i*(num-i));
            res = Math.max(res,v1*(num-i));
            res = Math.max(res,v2*i);
        }
        dp[num] = res;
        return res;
    }
    public int integerBreak(int n) {
        dp = new Integer[n+1];
        int res = compute(n);
        return res;
    }
}